File: Heap/Python/Path With Minimum Effort.py
# Problem link: https://leetcode.com/problems/path-with-minimum-effort/
# Time: O(m * n * log(m * n))
# Space: O(m * n)
class Solution:
    def minimumEffortPath(self, heights: List[List[int]]) -> int:
        rows, cols = len(heights), len(heights[0])
        minHeap, visit, directions = [(0, 0, 0)], set(), [[1, 0], [0, 1], [-1, 0], [0, -1]]


        while minHeap:
            t, r, c = heapq.heappop(minHeap)

            if (r, c) in visit:
                continue

            if r == rows - 1 and c == cols - 1:
                return t
            
            visit.add((r, c))

            for row, col in directions:
                if not(row + r < 0 or row + r == rows or col + c < 0 or col + c == cols):
                    time = max(t, abs(heights[r][c] - heights[row + r][col + c]))
                    heapq.heappush(minHeap, (time, row + r, col + c))
        
        return -1

File: Heap/Java/Path With Minimum Effort.java
// Problem link: 
// Time: O(m * n * log(m * n))
// Space: O(m * n)
class Solution {
    public int minimumEffortPath(int[][] heights) {
        int rows = heights.length, cols = heights[0].length, newRow, newCol, newTime, time, row, col;
        int[] triplet;
        
        PriorityQueue<int[]> minHeap = new PriorityQueue<>(Comparator.comparingInt(a -> a[0]));
        Set<String> visited = new HashSet<>();
        int[][] directions = {{1, 0}, {0, 1}, {-1, 0}, {0, -1}};
        
        minHeap.offer(new int[]{0, 0, 0});
        
        while (!minHeap.isEmpty()) {
            triplet = minHeap.poll();
            time = triplet[0];
            row = triplet[1];
            col = triplet[2];
            
            if(visited.contains(row + " " + col))
                continue;
            
            if(row == rows - 1 && col == cols - 1)
                return time;
            
            visited.add(row + " " + col);
            
            for(int[] dir : directions) {
                newRow = row + dir[0];
                newCol = col + dir[1];
                
                if(newRow >= 0 && newRow < rows && newCol >= 0 && newCol < cols) {
                    newTime = Math.max(time, Math.abs(heights[row][col] - heights[newRow][newCol]));
                    minHeap.offer(new int[]{newTime, newRow, newCol});
                }
            }
        }
        
        return -1;
    }
}
